W-21449660 new feature to manage dynamic client registration (#453)
mulesoft/docs-access-management: modules/ROOT/pages/configure-client-management-openid-task.adoc

= Configuring OpenID Connect Client Management

> Configure an external OpenID Connect (OIDC) identity provider (IdP) to handle client management for your Anypoint Platform APIs. Anypoint Platform supports IdPs that implement the OpenID Connect Dynamic Client Registration (DCR) standard; Salesforce, Okta, and OpenAM (v14 for OpenAM) are explicitly supported. This allows API portals to dynamically register client applications with your chosen external IdP when requesting access.

Reviewer: ```suggestion
Configure an external OpenID Connect (OIDC) identity provider (IdP) to handle client management for your Anypoint Platform APIs. Anypoint Platform supports IdPs that implement the OpenID Connect Dynamic Client Registration (DCR) standard, including Salesforce, Okta, and OpenAM (v14 for OpenAM). This allows API portals to dynamically register client applications with your chosen external IdP when requesting access.
```

Enables deletion of clients created with this integration.
* *Enable client deletion and updates in IdP*
+
To use this option, you must also select the *Enable client deletion in Anypoint Platform* option. This option enables you to update and delete external clients in the configured IdP through an outbound call made by Anypoint Platform to `{clientRegistrationUrl}/{clientID}`. The `clientRegistrationUrl` is the value you configured in *Client Registration URL*. For example: `+https://identity.example.com/oauth2/connect/register/6779ef20e75817b79602+`.
* *Use registration credentials for client management*
+

> Enables lifecycle management (update and delete) of dynamically registered client applications through Anypoint Platform. When enabled, Anypoint Platform automatically adapts to your IdP's capabilities. If the IdP returns `registration_access_token` and `registration_client_uri` (per RFC 7592), these credentials are used for direct client management. Otherwise, the existing mechanism is used.

Reviewer 2: I think we can remove this because we do update/delete of DCR clients even this option is not selected.

URL that provides the user’s identity, encoded in a secure JSON Web Token.
. Click *Create*.

> After this is successfully configured, you can apply the OpenID Connect OAuth Token Enforcement policy to your API Gateways through API Manager. Requesting API access through API portals dynamically generates client applications in the configured IdP that acts as a token provider.

Reviewer: ```suggestion
After this is successfully configured, you can use API Manager to apply the OpenID Connect OAuth Token Enforcement policy to your API Gateways. Requesting API access through API portals dynamically generates client applications in the configured IdP that acts as a token provider.
```

[NOTE]
For Okta, the Okta admin must assign the dynamically generated clients to a user or a group of users so they can receive access tokens by sending over the Client ID and Client Secret.

[[manage-dynamic-client-registration-and-lifecycle-updates]]
== Manage Dynamic Client Registration and Lifecycle Updates

Anypoint Platform supports lifecycle management (create, update, and delete) of dynamically registered OAuth2 clients. When you enable the *Use registration credentials for client management* option, Anypoint Platform automatically adapts to your IdP's capabilities:

> * If your IdP returns `registration_access_token` and `registration_client_uri` (as specified in RFC 7592), Anypoint Platform persists these credentials and uses them for direct client updates and deletions through the standard RFC 7592 endpoints.

Reviewer 2: Do we need to provide a link to RFC 7592?

Reviewer 2: https://www.rfc-editor.org/rfc/rfc7592.html

* If your IdP doesn't return these registration credentials, Anypoint Platform uses the existing client management mechanism, which allows you to intercept and handle update and delete requests according to your IdP's specific API requirements.

This adaptive approach ensures that client management works regardless of whether your IdP fully implements the RFC 7592 standard. Backward compatibility is maintained for existing configurations.

[[enable-client-management]]
=== Enable Client Management

> To enable lifecycle management for dynamically registered clients:

Reviewer 2: I'm not sure if we need this info. Even users don't select 'Use registration credentials for client management'. We still update and delete clients.

. Navigate to *Access Management* and click *Client Providers*.
. Select your configured client provider.
. Under the *Advanced settings* section, select *Use registration credentials for client management*.
. Click *Save*.

[IMPORTANT]

> Client management only applies to clients created after you enable this feature. Clients created before enabling this setting don't have the required registration data and can't be updated or deleted through Anypoint Platform.

Reviewer 2: I think we can re-word it to give more details: 

It is recommended that you create a new client provider if you want to use Use registration credentials for client management for consistency.

Before we this feature is available, we don't save registration_access_token and registration_client_uri. After this feature is available, we save the info. If you don't enable "Use registration credentials", we use the existing mechanism. If you enable it, we start using registration_access_token and registration_client_uri that may cause a change in behavior.

> Client management only applies to clients created *after* you enable this feature. Clients created before enabling this setting don't have the required registration data and can't be updated or deleted through Anypoint Platform.

Reviewer: ```suggestion
Client management only applies to clients created after you enable this feature. Clients created before enabling this setting don't have the required registration data and can't be updated or deleted through Anypoint Platform.
```

=== How Client Management Works

> When you enable client management, Anypoint Platform handles update and delete operations based on what your IdP provides during client registration.

Reviewer 2: User enable client management in IDP using the option below
"Enable client deletion and updates in IdP"

If the user also select "Use registration credentials", we use registration_access_token and registration_client_uri if IDP return them. If IDP don't return them, we use existing mechanism
